20_1 DataGrid calculateCustomSummary: docissues (#1816)
DevExpress/devextreme-documentation: api-reference/10 UI Components/dxDataGrid/1 Configuration/summary/calculateCustomSummary.md

The resulting summary item's value.

##### field(options.value): any

> If the **totalItems**.[column](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/#column) property is set, this field contains the value from this column. If **totalItems**.[showInColumn](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/#showInColumn) is used instead, this field contains all values from a row.

Reviewer: As I see, the information was already there, yet readers still don't notice it. Let's change the description more drastically then:

```suggestion
A column value used in calculation. If you need an entire data object in this field, do not specify the summary item's [column](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/#column) property. To display the summary in this case, use the [showInColumn](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/#showInColumn) property instead.
```

> A column value used in calculation. If you need an entire data object in this field, do not specify the summary item's [column](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/#column) property. To display the summary in this case, use the [showInColumn](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/#showInColumn) property instead.

Reviewer 2: "in calculation" In which calculation?

                                break;
                            case "finalize":
                                // Assigning the final value to "totalValue" here

>                                 break;

Reviewer: I vote for keeping the break. Sure, the code works without it, but it's good to show it just in case somebody wants to add a default case.

        if(options.name == "customSummary1") {
            switch(options.summaryProcess) {
                case "start":
                    // Initializing "totalValue" here
                    break;
                case "calculate":
                    // Modifying "totalValue" here
                    break;
                case "finalize":
                    // Assigning the final value to "totalValue" here
                    break;
            }
        }

        // Calculating "customSummary2"
        if(options.name == "customSummary2") {
            // ...
            // Same "switch" statement here
        }
    }

    function App() {
        return (
            <DataGrid ... >
                <Summary calculateCustomSummary={calculateCustomSummary}>
                    <TotalItem
                        name="customSummary1"
                        summaryType="custom"
                    />
                    <TotalItem
                        name="customSummary2"
                        summaryType="custom"
                    />
                </Summary>
            </DataGrid>
        );
    }

    export default App;

---

> You can use the **value** field to access a column value. If you do not specify a [column](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/summary/totalItems/#column) by which to calculate the summary, the **value** field contains an entire data object. However, this object misses values from unbound columns calculated in the [calculateCellValue](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#calculateCellValue) function. If you need these values for your custom summary, call the **calculateCellValue** function from inside **calculateCustomSummary**, as shown below. In this example, the calculateArea (calculateCellValue) function creates an unbound column 'Area'. The same function is called from the calculateAreaSummary (calculateCustomSummary) function to compute the sum of areas for selected rows.

Reviewer 2: - What do you mean when you say "to access"? To get the value? To retrieve the value?
- "by which to calculate the summary" Do you mean 'used to calculate the summary' or 'for which to calculate the summary'?

    import { NgModule } from '@angular/core'; 
    import { AppComponent } from './app.component'; 
    import { DxDataGridModule } from 'devextreme-angular'; 
    
    @NgModule({ 
        declarations: [ 
            AppComponent 
        ], 
        imports: [ 
            BrowserModule, 
            DxDataGridModule 
        ], 
        providers: [ ], 
        bootstrap: [AppComponent] 
    }) 
    
    export class AppModule { } 

##### Vue

    <!-- tab: App.vue -->
    <template>
        <DxDataGrid ... >
            <DxSummary :calculate-custom-summary="calculateCustomSummary">
                <DxTotalItem
                    name="сustomSummary1"
                    summary-type="custom"
                />
                <DxTotalItem
                    name="сustomSummary2"
                    summary-type="custom"
                />
            </DxSummary>
        </DxDataGrid>
    </template>
    <script>
    import { DxDataGrid, DxSummary, DxTotalItem } from 'devextreme-vue/data-grid';

    export default {
        components: {
            DxDataGrid,
            DxSummary, 
            DxTotalItem
        },
        methods: {

>             calculateArea(rowData) {
>                 return rowData.width * rowData.height;
>             },

Reviewer: ```suggestion
```

            columns: [
                "width", "height",
                {
                    dataField: "Area",
                    calculateCellValue: calculateArea
                }
            ],
            summary: {
                totalItems: [{
                    name: "AreaSummary",
                    summaryType: "custom"
                    showInColumn: "Area",
                    displayFormat: "Total Area: {0}",
                }],
                calculateCustomSummary: calculateAreaSummary
            }
        });
    });

##### Angular

    <!-- app.component.ts -->
    import { Component } from '@angular/core';
    // ...
    export class AppComponent {
        selectedRows: number[];

        constructor() {
            // ...
            this.selectedRows = [12, 13];
            this.calculateAreaSummary = this.calculateAreaSummary.bind(this);
        }

        calculateArea(rowData) {
            return rowData.width * rowData.height;
        }

        calculateAreaSummary(options) {
            if (options.name === "AreaSummary") {
                if (options.summaryProcess === "start") {
                    options.totalValue = 0;
                }

>                 if (options.component.isRowSelected(options.value.ID)) {

Reviewer: `if (options.summaryProcess === "calculate") {` is missing

> ---

Reviewer: Add Vue and React versions.

> **calculateCustomSummary** is called before [calculateCellValue](/Documentation/ApiReference/UI_Components/dxDataGrid/Configuration/columns/#calculateCellValue). If you want to create a new column and use its values in a custom summary, the following example demonstrates how to implement these functions:

Reviewer: - You give some facts, which I'm sure are true, but the use case isn't clear to me. The first sentence in a paragraph should present the main thought of this paragraph. So what's this thought?
- Elaborate on what is happening in the code sample.

---
##### jQuery

    <!--JavaScript-->

>     const  calculateArea = (rowData) => {

Reviewer: ```suggestion
    const calculateArea = (rowData) => {
```

>     const  calculateArea = (rowData) => {

Reviewer: Why is this an arrow function, but calculateAreaSummary below is declared as a regular function?

        return rowData.width * rowData.height;
    }

    function calculateAreaSummary(options) {

>         if (options.name === "calculateAreaSummary") {

Reviewer: ```suggestion
        if (options.name === "AreaSummary") {
```

Reviewer: Same mistake in samples for other platforms.

            if (options.summaryProcess === "start") {
                options.totalValue = 0;
            }
            if (options.summaryProcess === "calculate") {

>                     options.totalValue += calculateArea(options.value);

Reviewer: ```suggestion
                options.totalValue += calculateArea(options.value);
```

            }
        }
    }

    $(function(){    
        $("#gridContainer").dxDataGrid({
            // ...
            columns: [
                "width", "height",
                {
                    dataField: "Area",
                    calculateCellValue: calculateArea
                }
            ],
            summary: {
                totalItems: [{

>                         name: "AreaSummary",
>                         summaryType: "custom"
>                         showInColumn: "Area",
>                         displayFormat: "Total Area: {0}",
>                     }
>                 ],

Reviewer: ```suggestion
                    name: "AreaSummary",
                    summaryType: "custom"
                    showInColumn: "Area",
                    displayFormat: "Total Area: {0}",
                }],
```

                calculateCustomSummary: calculateAreaSummary
            }
        });
    });

##### Angular

>     <!--TypeScript-->

Reviewer: Use an updated template for code samples, the one where code is split across tabs named like files in a real project.

    import { Dx{WidgetName}Module } from "devextreme-angular";
    // ...
    export class AppComponent {

>         constructor(private service: Service) {

Reviewer: ```suggestion
        constructor() {
```

            // ...
            this.calculateAreaSummary = this.calculateAreaSummary.bind(this);
        }

        calculateArea(rowData) {
            return rowData.width * rowData.height;
        }

        calculateAreaSummary(options) {
            if (options.name === "calculateAreaSummary") {
                if (options.summaryProcess === "start") {
                    options.totalValue = 0;
                }
                if (options.summaryProcess === "calculate") {
                    options.totalValue += this.calculateArea(options.value);
                }
            }
        }
    }
    @NgModule({
        imports: [
            // ...

>             Dx{WidgetName}Module

Reviewer: Don't use placeholders if the description is not inherited.

        ],
        // ...
    })

    <!--HTML-->
    <dx-{widget-name} ... >
        <dxi-column dataField="width"></dxi-column>
        <dxi-column dataField="height"></dxi-column>
        <dxi-column
            dataField="Area"

>             [calculateCellValue]="calculateArea"
>         ></dxi-column>

Reviewer: ```suggestion
            [calculateCellValue]="calculateArea">
        </dxi-column>
```

> 

Reviewer: ```suggestion
```

        <dxo-summary [calculateCustomSummary]="calculateAreaSummary">
            <dxi-total-item
                name="AreaSummary"
                summaryType="custom"
                showInColumn="Area"

>                 displayFormat="Total Area: {0}"
>             >

Reviewer: ```suggestion
                displayFormat="Total Area: {0}">
```

    import { DxDataGrid, DxColumn, DxSummary, DxTotalItem } from 'devextreme-vue/data-grid';

    export default {
        components: {
            DxDataGrid,
            DxColumn,
            DxSummary, 
            DxTotalItem
        },
        data() {
            return {
                // ...
                calculateArea(rowData) {
                    return rowData.width * rowData.height;
                },
                calculateAreaSummary(options) {
                    if (options.name === "calculateAreaSummary") {
                        if (options.summaryProcess === "start") {
                            options.totalValue = 0;
                        }
                        if (options.summaryProcess === "calculate") {
                            options.totalValue += this.calculateArea(options.value);
                        }
                    }
                },
            };
        },
    }
    </script>

##### React

    <!-- tab: App.js -->
    import React from 'react';
    import DataGrid, { Column, Summary, TotalItem } from 'devextreme-react/data-grid';

    function calculateArea(rowData) {
        return rowData.width * rowData.height;

>     };

Reviewer: ```suggestion
    }
```

    function calculateAreaSummary(options) {
        if (options.name === "calculateAreaSummary") {
            if (options.summaryProcess === "start") {
                options.totalValue = 0;
            }
            if (options.summaryProcess === "calculate") {
                options.totalValue += calculateArea(options.value);
            }
        }
    }

    function App() {
        // ...
        return (

>             <DataGrid ...>

Reviewer: ```suggestion
            <DataGrid ... >
```

                <Column dataField="width" />
                <Column dataField="height" />
                <Column dataField="Area" calculateCellValue={calculateArea} />
                <Summary calculateCustomSummary={calculateAreaSummary}>
                    <TotalItem
                        name="AreaSummary"
                        summaryType="custom"
                        showInColumn="Area"
                        displayFormat="Total Area: {0}"
                    />
                </Summary>
            </DataGrid>
        );
    }

    export default App;

---

> Note that the example above uses totalItems.**showInColumn** instead of totalItems.**column**. This is because **calculateCustomSummary** works with values from multiple columns.

Reviewer: The effect and the cause are switched places. We need entire data objects within calculateCustomSummary. To achieve that, we don't specify the column. Not the other way around. Give this information above the example.
